rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,117 @@
+* (-) It requires significant additional effort to implement and maintain an appropriate system to make use of the version information.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Never introduce backwards-incompatible changes to the language.
+
+  * (+) No additional effort required.
+  * (-) Requires additions to be made very carefully.
+  * (-) Makes solving some well-known problems impossible.
+
+- Use the output of [`builtins.langVersion`] for specifying the version of the Nix language.
+
+  * (+) This would serve other Nix language evaluators which are not and should not be tied to the rest of Nix.
+  * (-) `builtins.langVersion` is currently only internal and undocumented.
+    * (+) Documentation is easy to add.
+      * (-) Requires adding another built-in to the public API.
+    * (-) Using a language feature requires an additional steps from users to determine the current version.
+      * (+) We can add a command line option such that it is not more effort than `nix --version`.
+        * (-) Requires adding another command line option to the public API.
+  * (+) The Nix language version is decoupled Nix version numbering.
+    * (+) It changes less often than the Nix version.
+      * (-) That was probably due to making changes being so hard.
+    * (-) There are two version numbers to keep track of.
+  * (-) The magic comment should reflect that it's specifying the *Nix language* version, which would make it longer.
+    * (-) Renaming the Nix language will be impossible once the mechanism is part of stable Nix.
+
+[`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957
+
+- Use a magic string that is incompatible with evaluators prior to the feature, e.g. `%? Nix <version>`.
+
+  * (+) Makes clear that the file is not intended to be used without explicit handling of compatibility.
+    * (-) Cannot be introduced gradually.
+  * (+) Such a breaking change could also be reserved for later iterations of the Nix language.
+
+# Prior art
+
+- [Rust `edition` field]
+
+  Rust has an easier problem to solve. Cargo files are written in TOML, so the `edition` information does not have to be part of Rust itself.
+
+- [Flakes `edition` field]
+
+  There had been an attempt to include an `edition` field into the Flakes schema.
+  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.
+
+[Rust `edition` field]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field
+[Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Is the proposed magic number already in use in [other file formats](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)?
+- Should we allow multiple known-good versions in one line?
+
```

Comment:
How does this RFC impact the development of language server, formatters, parsers of the Nix language, and other tools that depend on being able to understand the Nix language?

Right now and after 20 years we don't yet have a perfect tool for any of the tasks mentioned previously, and I worry that introducing more language versions will make that task even harder

However, there is something I find very interesting about this RFC: the fact that we can finally create a Nix language version that can be parsed 100% correctly without using GNU Bison, the GLR algorithm, Flex (lexer with states), or tools/algorithms not present in more modern programming languages like Rust

Reply:
> not present in more modern programming languages like Rust

What do you mean? `rnix-parser` for example has compatibility tests with C++ Nix that guarantee that ASTs are parsed identically (in the success case, there are some minor discrepancies in error cases).

Reply:
Let me clarify: there is no equivalent of GNU Bison in Rust, there is no equivalent of Flex in Rust, and there is no tool that can run the GLR algorithm in Rust. The same applies to other programming languages... GNU Bison, Flex and GLR are a particularly hard combination to port beyond c++. Yes, rnix-parser, nixel, tree-sitter-nix, and others achieved this task to some extent, but I think their authors would agree this is not an easy task, even less easy to do if we had many versions, and yes you can have all the tests and fuzzing in the world, and spend uncountable hours trying to get it correct, yet it would be nice if the Nix language just were simple to port out of the box (for instance by using a simple lexer with no states and a common parsing algorithm like LR)

What I try to express with this comment is that something I find interesting in allowing the language to evolve is making the Nix language possible to parse with simple tools and algorithms, allowing ports, easily available in all (especially more recent) programming languages. 

I'm not attacking or judging any of the existing implementations, just mentioning that this RFC can potentially allow for changing the language in a way that makes the proliferation of more implementations easier, which benefits the community

Reply:
> we can finally create a Nix language version that can be parsed 100% correctly without using [obscure tools]

Not sure how this RFC addresses this. Right now it would give some hints about the grammar to use for parsing a given expression.

Reply:
yes, this RFC does not address this, and it shouldn't

maybe we can look at my comment from the perspective that the following two tradeoffs can cancel out together:
1. having more versions of the nix language makes it harder to create tools around it
1. making the nix language as easy to port as possible (hopefully just copy-pasting the parser generator), then it would be easier for those tools to keep up with the changes

if we don't have "2" or another mechanism, then we should not overlook "1"

another mechanisms for "2" could be making the official Nix parser "open" (in the sense that it's easy to reuse in other projects), same as #134 tries to have a standalone nix-store, having a standalone nix-parse that language servers, formatters, and other tools can use would cancel out the problem in "1"

Reply:
> making the nix language as easy to port as possible (hopefully just copy-pasting the parser generator)

The syntax (and parsing thereof) is definitely a big concern in the semantics of Nix, but it is nowhere near the biggest. You would have to make many extremely invasive (not just cosmetic) changes to make Nix as simple as "just copy-pasting the parser generator", and I'm not sure how much appetite for that there is.

> having a standalone nix-parse that language servers, formatters, and other tools can use

We do! It's `rnix-parser`. It parses all correct code correctly. There's literally only one known issue, and it's about code that isn't valid anyways.

Reply:
> there is no tool that can run the GLR algorithm in Rust.

This is [not](https://tree-sitter.github.io/tree-sitter/creating-parsers#:~:text=Tree-sitter is based,algorithm) [true](https://github.com/tree-sitter/tree-sitter/blob/master/docs/section-3-creating-parsers.md?plain=1#L261).

Reply:
> there is no tool that can run the GLR algorithm in Rust. ... GLR are a particularly hard combination to port beyond c++.

This is true only in a very misleading sense.

Tree-sitter [implements the GLR algorithm](https://github.com/tree-sitter/tree-sitter/blob/9737fcb558b72683e81286fa46457e48d518c672/docs/section-3-creating-parsers.md?plain=1#L261) and the GLR "frontend" (parse table generator) is written in Rust.  The parsing code it emits, which implements the other half of the GLR algorithm (graph-structured stack) is C code.

It very easy to emit Rust code instead.  [It has already been done](https://github.com/tree-sitter/tree-sitter/pull/638).  The tree-sitter devs have [consciously chose not to do that](https://github.com/tree-sitter/tree-sitter/issues/1670) since it would reduce the applicability of tree-sitter (many of its users want a C-only runtime).  To be fair, it's not entirely clear if Rust's safety guarantees are a big deal here.  This isn't human-written code, so it has the same safety auditing requirements as `rustc`'s own backend.

So the reason why there are no popular all-Rust implmenentations of the "backend" of the GLR algorithm has very little to do with Rust's suitability.  It's mostly a result of there being an extremely good implementation that emits C code, and that emitter being pretty easy to audit for safety (unlike hand-written C code).